File: Book/again/08-4. 효율적인 화폐 구성.js
/**
 * target을 낼 때 필요한 최소 화폐 갯수 출력
 * 불가능하다면 -1 출력
 */

const target = 15;
const coin = [2, 3];
const target2 = 4;
const coin2 = [3, 5, 7];

function solution(target, coin) {
  const INF = 987654321;
  const array = Array(target + 1).fill(INF);
  // 있는 동전은 1로 만든다.
  for (let i = 0; i < coin.length; i++) {
    const current = coin[i];
    array[current] = 1;
  }
  coin.sort((a, b) => a - b);
  for (let i = coin[0]; i <= target; i++) {
    for (let j = 0; j < coin.length; j++) {
      if (i - coin[j] >= 0) {
        if (array[i - coin[j]] !== INF)
          array[i] = Math.min(array[i], array[i - coin[j]] + 1);
      }
    }
  }
  console.log(array);
  return array[target] !== INF ? array[target] : -1;
}

console.log(solution(target2, coin2));

File: Book/again/08-4. 효율적인 화폐 구성.py
n, m = map(int, input().split())
coin = []
for _ in range(n):
    coin.append(int(input()))

array = [10001] * (m + 1)
array[0] = 0
for i in range(n):
    for j in range(coin[i], m + 1):
        if array[j - coin[i]] != 10001:
            array[j] = min(array[j], array[j - coin[i]] + 1)

print(array[m] if array[m] != 10001 else -1)
